compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -1654,7 +1655,31 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                 match *cast_kind {
                     CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer, coercion_source) => {
                         let is_implicit_coercion = coercion_source == CoercionSource::Implicit;
-                        let src_sig = op.ty(body, tcx).fn_sig(tcx);
+                        let src_ty = op.ty(body, tcx);
+                        let mut src_sig = src_ty.fn_sig(tcx);
+                        if let ty::FnDef(def_id, _) = src_ty.kind() {
+                            // If we are casting a safe target feature function to a safe fn
+                            // pointer, we need to mark the function as safe explicitly.
+                            if tcx.codegen_fn_attrs(def_id).safe_target_features
+                                && let ty::FnPtr(_, target_hdr) = *ty.kind()
+                                && target_hdr.safety.is_safe()
+                            {
+                                let coercee_features =
+                                    &tcx.codegen_fn_attrs(def_id).target_features;
+                                let body_features =
+                                    &tcx.codegen_fn_attrs(body.source.def_id()).target_features;
+                                if is_target_feature_call_safe(
+                                    tcx,
+                                    &coercee_features,
+                                    &body_features,
+                                ) {
+                                    src_sig = src_sig.map_bound(|sig| ty::FnSig {
+                                        safety: hir::Safety::Safe,
+                                        ..sig
+                                    })
+                                }
+                            }
```

Comment:
Yea, as you suggested in DMs, I think you should move this out into a TyCtxt method and reuse it in typeck.